python/ray/data/_internal/planner/exchange/sort_task_spec.py

```diff
@@ -174,34 +174,28 @@ def sample_boundaries(
         # TODO(zhilong): Update sort sample bar before finished.
         samples = sample_bar.fetch_until_complete(sample_results)
         del sample_results
-        samples = [s for s in samples if len(s) > 0]
+        samples: List[Block] = [s for s in samples if len(s) > 0]
         # The dataset is empty
         if len(samples) == 0:
             return [None] * (num_reducers - 1)
+
+        # Convert samples to a sorted list[tuple[...]] where each tuple represents a
+        # sample.
+        # TODO: Once we deprecate pandas blocks, we can avoid this conversion and
+        # directly sort the samples.
         builder = DelegatingBlockBuilder()
         for sample in samples:
             builder.add_block(sample)
         samples = builder.build()
+        samples = BlockAccessor.for_block(samples).to_numpy(columns=columns)
+        samples = sorted(zip(*samples.values()))
```

Comment:
Don't overwrite same var with values of different types -- this confuses the type inference and makes it harder to reason about the code (instead, treat it as staticly-typed language)

Reply:
Let's also leave a comment that `zip` is critical here to do transposition (from list of column vals to list of tuples)